nifi-nar-bundles/nifi-salesforce-bundle/nifi-salesforce-processors/src/main/java/org/apache/nifi/processors/salesforce/PutSalesforceObject.java

```diff
@@ -146,69 +136,85 @@ public void onTrigger(ProcessContext context, ProcessSession session) throws Pro
-        try {
         try (InputStream in = session.read(flowFile);
              RecordReader reader = readerFactory.createRecordReader(flowFile, in, getLogger());
              ByteArrayOutputStream out = new ByteArrayOutputStream();
-             WriteJsonResult writer = getWriter(extender = new RecordExtender(reader.getSchema()), out)) {
+             WriteJsonResult writer = getWriter(recordExtender = getExtender(reader), out)) {
 
-            int count = 0;
             Record record;
-
             while ((record = reader.nextRecord()) != null) {
                 count++;
                 if (!writer.isActiveRecordSet()) {
                     writer.beginRecordSet();
                 }
 
-                MapRecord extendedRecord = extender.getExtendedRecord(objectType, count, record);
+                MapRecord extendedRecord = recordExtender.getExtendedRecord(objectType, count, record);
                 writer.write(extendedRecord);
 
                 if (count == maxRecordCount) {
                     count = 0;
-                    processRecords(objectType, out, writer, extender);
+                    postRecordBatch(objectType, out, writer, recordExtender);
                     out.reset();
                 }
             }
             if (writer.isActiveRecordSet()) {
-                processRecords(objectType, out, writer, extender);
+                postRecordBatch(objectType, out, writer, recordExtender);
             }
-          }
-          session.transfer(flowFile, REL_SUCCESS);
-          long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
-          session.getProvenanceReporter().send(flowFile, salesforceRestService.getVersionedBaseUrl()+ "/composite/tree/" + objectType, transferMillis);
-        } catch (MalformedRecordException e) {
-            getLogger().error("Couldn't read records from input", e);
-            transferToFailure(session, flowFile, e);
-        } catch (SchemaNotFoundException e) {
-            getLogger().error("Couldn't create record writer", e);
-            transferToFailure(session, flowFile, e);
-        } catch (Exception e) {
-            getLogger().error("Failed to put records to Salesforce.", e);
-            transferToFailure(session, flowFile, e);
         }
     }
 
+    private SalesforceConfiguration createSalesforceConfiguration(ProcessContext context) {
+        String salesforceVersion = context.getProperty(API_VERSION).getValue();
+        String instanceUrl = context.getProperty(SALESFORCE_INSTANCE_URL).getValue();
+        OAuth2AccessTokenProvider accessTokenProvider =
+                context.getProperty(TOKEN_PROVIDER).asControllerService(OAuth2AccessTokenProvider.class);
+        return SalesforceConfiguration.create(instanceUrl, salesforceVersion,
+                () -> accessTokenProvider.getAccessDetails().getAccessToken(), 0);
+    }
+
+    private void handleInvalidFlowFile(ProcessSession session, FlowFile flowFile) {
```

Comment:
The handleInvalidFlowFile and transferToFailure methods are quite similar.
You can change transferToFailure a bit and reuse it in handleInvalidFlowFile.
(The transferToFailure uses the message from the exception parameter only.)

 private void transferToFailure(ProcessSession session, FlowFile flowFile, String message) {
        flowFile = session.putAttribute(flowFile, ATTR_ERROR_MESSAGE, message);
        session.transfer(session.penalize(flowFile), REL_FAILURE);
    }

 private void handleInvalidFlowFile(ProcessSession session, FlowFile flowFile) {
        getLogger().error("Salesforce object type not found among the incoming FlowFile attributes");
        transferToFailure(session, flowFile, "Salesforce object type not found among FlowFile attributes");
    }